src/routes/spsp.ts

```diff
@@ -0,0 +1,76 @@
+import { randomBytes, generateReceiptSecret } from '../util/crypto'
+import { IncomingMessage, ServerResponse } from 'http'
+import fetch from 'node-fetch'
+import * as Koa from 'koa'
+import * as Router from 'koa-router'
+import * as url from 'url'
+
+function resolvePointer (pointer: string): string {
+  if (!pointer.startsWith('$')) {
+    return pointer
+  }
+
+  const protocol = (process.env.NODE_ENV === 'test') ? 'http://' : 'https://'
+  const url = new URL(protocol + pointer.substring(1))
+  if (url.pathname === '/') {
+    url.pathname = '/.well-known/pay'
+  }
+
+  return url.href
+}
+
+export const router = new Router()
+
+router.get('/(.*)', async (ctx: Koa.Context) => {
+  if (ctx.accepts('application/spsp4+json') && ctx.req.url) {
+    let spspEndpoint: string
+    if (ctx.config.spspEndpointsUrl) {
+      const id = encodeURIComponent(ctx.req.url.substring(1))
+      const endpointsRes = await fetch(`${ctx.config.spspEndpointsUrl}?id=${id}`)
+      if (endpointsRes.status !== 200) {
+        console.error(await endpointsRes.text())
+        ctx.throw(404)
+      }
+      spspEndpoint = await endpointsRes.text()
+    } else {
+      spspEndpoint = decodeURIComponent(ctx.req.url.substring(1))
+    }
+    const nonce = randomBytes(16)
+    const secret = generateReceiptSecret(ctx.config.receiptSeed, nonce)
+    ctx.proxyServer.once('proxyRes', (proxyRes: IncomingMessage, req: IncomingMessage, res: ServerResponse) => {
```

Comment:
Won't this event handler (sometimes) mismatch with the `ctx` when multiple requests are in-flight? I'm not 100% sure, but it looks like the `proxyServer` object is shared between all middleware chains.

Reply:
:heavy_check_mark: you're right. Thanks for catching that. (I think the issue predates this pr.)
https://github.com/coilhq/receipt-verifier/pull/23/commits/77d1257997783c0a80125c2fc2c5aae209c2f1e3 swaps out `http-proxy` for an actual koa middleware.